tracks/python/exercises/isbn-verifier/mentoring.md

```diff
@@ -0,0 +1,111 @@
+# Mentoring
+
+## Problem and Challenges
+
+The problem asks you to write an ISBN validation checker.
+
+* Any "-" in the input string should be ignored.
+* A valid ISBN must be 10 values long.
+* A valid ISBN must be entirely digits, other than the last character.
+* A valid ISBN may have `"X"` as the last character, which should be treated as a `10` value.
+* The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.
+
+## Reasonable Solution
+
+```python
+def is_valid(isbn: str) -> bool:
+    values = [value for value in isbn if value != "-"]
+
+    if len(values) != 10:
+        return False
+
+    if values[-1] == "X":
+        values[-1] = "10"
+
+    if any(not value.isdigit() for value in values):
+        return False
+
+    total = sum(int(value) * (10 - position) for position, value in enumerate(values))
+    return total % 11 == 0
+```
+
+## Talking Points
+
+### Generator expressions
+
+The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].
+
+```python
+total = 0
+for position, value in enumerate(values)):
+    total += int(value) * (10 - position)
+```
+
+Generator expressions are considered very Pythonic and enable compact and clean iteration.
+They are also very efficient.
```

Comment:
I could run benchmarks and dump values.
I could link to PyCon talks.
But this is guidance for mentors, like a research paper :)

Are you referring specifically to line 45? Or the entire argument in favor of using generators?

Reply:
Specifically "They are also very efficient."  This could be "This is known to be more efficient than other potential approaches." Maybe.  But if there are references it is good to provide further reading to support this statement.  Otherwise we could pose it as a question for exploration.

Reply:
Couching it in more vague language shifts it from a "claim from authority" to an "appeal to the people". You're saying the same thing ... but saying "it's known" vs "I know". It doesn't change the substance in any way.

Generators store less state than iterators and require fewer lookups, which is why they're faster than iterators. Generators don't build a list (which requires list appends and storing all the elements) which is why they are faster than building a list. Should I include these details?

There is PEP 255 where generators are introduced; the PEP lays out the reason for generators which includes the inefficiencies of lists and iterators for this use case. Generators are introduced in part *because* they are more efficient. I suppose the PEP itself is a pretty good reference for their efficiency, if someone wanted to dig through it. (See Motivation, paragraphs 3 and 4.)

Or I could just state they are more efficient ;)\

> Otherwise we could pose it as a question for exploration

Are we posing the question to mentors? Suggesting mentors suggest students benchmark it?

References:

https://www.dabeaz.com/generators/ and his PyCon talks are an amazing watch/listen on generators. Googling for "python generators vs iterators" produces numerous results. https://www.python.org/dev/peps/pep-0255/ motivation.

See https://medium.com/analytics-vidhya/python-iterators-vs-iterables-vs-generators-e15b9403292e

> Generator is efficient for writing fast and compact code. This is an advantage over Python iterators. They are also simpler to code than do custom iterator.

Reply:
So I think you are saying that you could reference PEP 255 where efficiencies are mentioned.  (More efficient in what?  CPU ticks, Memory, Power, Time complexity?) Or perhaps the references you show here are good items to show in the mentor notes?

> Are we posing the question to mentors?

This document can supply questions to pose to students, sometimes as copy/paste-able material.  And usually I avoid benchmark suggestions, unless the benchmarks are extremely simple, or if the student brings it up, and even them usually we find out if that is really the bottleneck in their solution, often by profiling instead, or at least before we benchmark.

Reply:
Reference to PEP-255 added.